test/AxelarGatewayUpgrade.js

```diff
@@ -0,0 +1,180 @@
+    });
+
+    it('should upgrade AxelarGateway through InterchainGovernance proposal', async () => {
+        const commandID = 0;
+        const target = gateway.address;
+        const nativeValue = 0;
+        const timeDelay = isHardhat ? 12 * 60 * 60 : 20;
+
+        const targetInterface = new Interface(gateway.interface.fragments);
+        const newGatewayImplementation = await gatewayFactory.deploy(auth.address, tokenDeployer.address).then((d) => d.deployed());
+        const newGatewayImplementationCodeHash = await getBytecodeHash(newGatewayImplementation);
+        const setupParams = '0x';
+        const calldata = targetInterface.encodeFunctionData('upgrade', [
+            newGatewayImplementation.address,
+            newGatewayImplementationCodeHash,
+            setupParams,
+        ]);
+
+        const [payload, proposalHash, eta] = await getPayloadAndProposalHash(commandID, target, nativeValue, calldata, timeDelay);
+
+        const payloadHash = solidityKeccak256(['bytes'], [payload]);
+        const commandIdGateway = getRandomID();
+        const sourceTxHash = keccak256('0x123abc123abc');
+        const sourceEventIndex = 17;
+
+        const approveData = buildCommandBatch(
+            await getChainId(),
+            [commandIdGateway],
+            ['approveContractCall'],
+            [
+                getApproveContractCall(
+                    governanceChain,
+                    governanceAddress,
+                    interchainGovernance.address,
+                    payloadHash,
+                    sourceTxHash,
+                    sourceEventIndex,
+                ),
+            ],
+        );
+
+        const approveInput = await getSignedWeightedExecuteInput(
+            approveData,
+            operators,
+            getWeights(operators),
+            threshold,
+            operators.slice(0, threshold),
+        );
+
+        await expect(gateway.execute(approveInput, getGasOptions()))
+            .to.emit(gateway, 'ContractCallApproved')
+            .withArgs(
+                commandIdGateway,
+                governanceChain,
+                governanceAddress,
+                interchainGovernance.address,
+                payloadHash,
+                sourceTxHash,
+                sourceEventIndex,
+            );
+
+        await expect(interchainGovernance.execute(commandIdGateway, governanceChain, governanceAddress, payload, getGasOptions()))
+            .to.emit(interchainGovernance, 'ProposalScheduled')
+            .withArgs(proposalHash, target, calldata, nativeValue, eta);
+
+        await waitFor(timeDelay);
+
+        const tx = await interchainGovernance.executeProposal(target, calldata, nativeValue);
+        const executionTimestamp = (await ethers.provider.getBlock(tx.blockNumber)).timestamp;
+
+        await expect(tx)
+            .to.emit(interchainGovernance, 'ProposalExecuted')
+            .withArgs(proposalHash, target, calldata, nativeValue, executionTimestamp)
```

Comment:
This test occasionally fails on live testnets where the calculated executionTimestamp differs by a few seconds from the timestamp emitted by the event. Not sure how to get around this.

Reply:
can you use null or omit the timestamp arg from withArgs?

Reply:
I can't emit an arg or put null, a possible work around would be something ugly like checking each event arg from the receipts independently. Something like this:
` const receipt = await tx.wait();
        const event = receipt.events?.find((e) => e.event === 'ProposalExecuted' && e.address === interchainGovernance.address);

        expect(event).to.not.be.undefined;
        expect(event.args.proposalHash).to.equal(proposalHash);
        expect(event.args.target).to.equal(target);
        expect(event.args.callData).to.equal(calldata);
        expect(event.args.value).to.equal(nativeValue);

        if (isHardhat) {
            expect(event.args.timestamp).to.equal(executionTimestamp);
        }` 
        Not sure if we want to to go down this route though.

Reply:
hmm, i'll play with it a bit. it's weird that timestamp retrieved from block is different than the event

Reply:
sounds good, let me know